Client/Frontend/Library/HistoryPanel/HistoryPanelViewModel.swift

```diff
@@ -0,0 +1,221 @@
+    var isFetchInProgress = false
+    var groupedSites = DateGroupedTableData<Site>()
+    
+    private var hasRecentlyClosed: Bool {
+        return profile.recentlyClosedTabs.tabs.count > 0
+    }
+    
+    let historyPanelNotifications = [ Notification.Name.FirefoxAccountChanged,
+                                      Notification.Name.PrivateDataClearedHistory,
+                                      Notification.Name.DynamicFontChanged,
+                                      Notification.Name.DatabaseWasReopened,
+                                      Notification.Name.OpenClearRecentHistory ]
+    
+    enum Sections: Int, CaseIterable {
+        case additionalHistoryActions
+        case today
+        case yesterday
+        case lastWeek
+        case lastMonth
+        case older
+
+        var title: String? {
+            switch self {
+            case .today:
+                return .LibraryPanel.Sections.Today
+            case .yesterday:
+                return .LibraryPanel.Sections.Yesterday
+            case .lastWeek:
+                return .LibraryPanel.Sections.LastWeek
+            case .lastMonth:
+                return .LibraryPanel.Sections.LastMonth
+            case .older:
+                return .LibraryPanel.Sections.Older
+            default:
+                return nil
+            }
+        }
+    }
+    
+    // MARK: - Inits
+    
+    init(profile: Profile) {
+        self.profile = profile
+    }
+    
+    deinit {
+        browserLog.debug("HistoryPanelViewModel DEinited.s")
+    }
+    
+    // MARK: - Lifecycles
+    
+    func viewDidLoad() {
+        reloadData()
+    }
+    
+    // MARK: - Private helpers
+    
+    /// Begin the process of fetching history data, and creating ASGroups from them. A prefetch also triggers this.
+    func reloadData() {
+        // Can be called while app backgrounded and the db closed, don't try to reload the data source in this case
+        guard !profile.isShutdown, !isFetchInProgress else {
+            browserLog.debug("HistoryPanel tableView data could NOT be reloaded! Either the profile wasn't shut down, or there's a fetch in progress.")
+            return
+        }
+        
+        fetchData().uponQueue(.global(qos: .userInteractive)) { result in
+            if let sites = result.successValue {
+                let fetchedSites = sites.asArray()
+                
+                self.populateHistorySites(fetchedSites: fetchedSites)
+                
+                if self.featureFlags.isFeatureActiveForBuild(.historyGroups) {
+                    self.populateASGroups()
+                }
+                
+                self.visibleSections = Sections.allCases.filter { section in
+                    self.groupedSites.numberOfItemsForSection(section.rawValue - 1) > 0
+                }
+            }
+        }
+    }
+    
+    /// A helper for the reload function.
+    private func fetchData() -> Deferred<Maybe<Cursor<Site>>> {
```

Comment:
Deferred. Yuck! haha

Reply:
I did a small attempt at async and well ... I think I need to read more docs 🥲